### PR TITLE
[jiterator] logical_and: complex

### PR DESCRIPTION
**Pull request by khushi-411**
Hi!
The PR follows: #74748

cc: @kshitij12345! :)

### CODE REVIEW

**Comment by kshitij12345** [comment]
Failure looks real:
```
======================================================================
[15291](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15291)
FAIL [0.281s]: test_dtypes_logical_and_cuda (__main__.TestCommonCUDA)
[15292](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15292)
----------------------------------------------------------------------
[15293](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15293)
Traceback (most recent call last):
[15294](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15294)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
[15295](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15295)
    method(*args, **kwargs)
[15296](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15296)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
[15297](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15297)
    result = test(self, **param_kwargs)
[15298](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15298)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 780, in test_wrapper
[15299](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15299)
    return test(*args, **kwargs)
[15300](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15300)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 815, in dep_fn
[15301](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15301)
    return fn(slf, *args, **kwargs)
[15302](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15302)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 973, in only_fn
[15303](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15303)
    return fn(self, *args, **kwargs)
[15304](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15304)
  File "test_ops.py", line 161, in test_dtypes
[15305](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15305)
    self.assertEqual(supported_dtypes, claimed_supported, msg=msg)
[15306](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15306)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2193, in assertEqual
[15307](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15307)
    msg=msg,
[15308](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15308)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1084, in assert_equal
[15309](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15309)
    raise error_metas[0].to_error()
[15310](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15310)
AssertionError: The supported dtypes for logical_and on cuda according to its OpInfo are
[15311](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15311)
        {torch.float32, torch.int8, torch.float64, torch.int16, torch.int32, torch.complex64, torch.int64, torch.bfloat16, torch.complex128, torch.float16, torch.uint8, torch.bool}, but the detected supported dtypes are {torch.float32, torch.int8, torch.float64, torch.int16, torch.int32, torch.int64, torch.bfloat16, torch.float16, torch.uint8, torch.bool}.
[15312](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15312)
        The following dtypes should be removed from the OpInfo: {torch.complex128, torch.complex64}.
[15313](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15313)

[15314](https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:15314)
----------------------------------------------------------------------
```

Looks like JITERATOR code fails to compile due to `&&` operator not being available for std::complex (you might have to add it, see https://github.com/pytorch/pytorch/pull/74161) 
https://github.com/pytorch/pytorch/runs/5748298918?check_suite_focus=true#step:9:13490

**Comment by khushi-411** [comment]
Got that! It looks interesting.
I'll update it soon. Thanks, @kshitij12345! :)

File: aten/src/ATen/cuda/llvm_complex.cpp [hunk 1]
```diff
@@ -477,6 +477,14 @@ operator!=(const _Tp& __x, const complex<_Tp>& __y)
     return !(__x == __y);
 }
 
+template<class _Tp>
+inline constexpr
+bool
+operator&&(const _Tp& __x, const complex<_Tp>& __y)
```
Comment (kshitij12345) [inline aten/src/ATen/cuda/llvm_complex.cpp:483]: I think you will have to add an overload for  `bool operator&&(complex<T>&, complex<T>&)`

Comment (kshitij12345) [reply aten/src/ATen/cuda/llvm_complex.cpp:483]: Ref : https://github.com/pytorch/pytorch/runs/5789888945?check_suite_focus=true#step:9:20384

Comment (khushi-411) [reply aten/src/ATen/cuda/llvm_complex.cpp:499]: Hi, @kshitij12345!
Thanks for your input! :)

I've added it. Looking into the other operators. I considered the situation in three different ways: both the inputs are complex, or if either args are complex.

Is this the right way? Thanks!

File: aten/src/ATen/cuda/llvm_complex.cpp [hunk 2]
```diff
@@ -477,6 +477,14 @@ operator!=(const _Tp& __x, const complex<_Tp>& __y)
     return !(__x == __y);
 }
 
+template<class _Tp>
+inline constexpr
+bool
+operator&&(const complex<_Tp>& __x, const complex<_Tp>& __y)
+{
+    return (__x.real() || __y.imag()) && (__x.imag() || __y.real());
```
Comment (kshitij12345) [inline aten/src/ATen/cuda/llvm_complex.cpp:485]: Shouldn't this be 

```c++
return (__x.real() || __x.imag()) && (__y.real() || __y.imag());
```

Comment (khushi-411) [reply aten/src/ATen/cuda/llvm_complex.cpp]: Yus! done :) Thanks!

**Comment by khushi-411** [comment]
Hi, @kshitij12345! Hi, @anjali411!
The CI turned green. Gentle ping to you both :)

**Comment by khushi-411** [comment]
Hi @kshitij12345, @anjali411! :)
Gentle ping!

**Review by anjali411 — APPROVED** [review]
LGTM. will wait for @kshitij12345 to take a final look

**Review by kshitij12345 — APPROVED** [review]
LGTM! Thanks @khushi-411

cc: @anjali411

**Pull request closed without merging**